File: Leetcode_POTD_2025/7_JULY/3480_Maximize_Subarrays_After_Removing_One_Conflicting_Pair.java
import java.util.ArrayList;
import java.util.List;
import java.util.Collections; // For max/min operations, if not using Math.max/min

class Solution {
    public long maxSubarrays(int n, int[][] conflictingPairs) {
        // Pre-process conflicting pairs: map right endpoint to list of left endpoints
        List<List<Integer>> mysticSeals = new ArrayList<>();
        for (int i = 0; i <= n; i++) {
            mysticSeals.add(new ArrayList<>());
        }

        for (int[] sigil : conflictingPairs) {
            mysticSeals.get(Math.max(sigil[0], sigil[1])).add(Math.min(sigil[0], sigil[1]));
        }

        long totalLoreFragments = 0;
        long primeInhibitor = 0;
        long secondaryInhibitor = 0;

        long[] liberationScrolls = new long[n + 1]; // Using array for direct indexing

        // Sweep line from left to right (1 to n)
        for (int scryingPoint = 1; scryingPoint <= n; scryingPoint++) {
            for (int forbiddenOrigin : mysticSeals.get(scryingPoint)) {
                if (forbiddenOrigin > primeInhibitor) {
                    secondaryInhibitor = primeInhibitor;
                    primeInhibitor = forbiddenOrigin;
                } else if (forbiddenOrigin > secondaryInhibitor) {
                    secondaryInhibitor = forbiddenOrigin;
                }
            }
            
            // Calculate baseline valid subarrays for current scryingPoint
            totalLoreFragments += scryingPoint - primeInhibitor;
            
            // Accumulate potential gain if the primary inhibitor were removed
            if (primeInhibitor > 0) {
                liberationScrolls[(int)primeInhibitor] += primeInhibitor - secondaryInhibitor;
            }
        }

        // Find the maximum possible gain from removing one conflicting pair
        long maximumInsight = 0;
        for (long scroll : liberationScrolls) {
            maximumInsight = Math.max(maximumInsight, scroll);
        }

        // Return total baseline + maximum gain
        return totalLoreFragments + maximumInsight;
    }
}

File: Leetcode_POTD_2025/7_JULY/3480_Maximize_Subarrays_After_Removing_One_Conflicting_Pair.py
class Solution(object):
    def maxSubarrays(self, n, conflictingPairs):
        """
        :type n: int
        :type conflictingPairs: List[List[int]]
        :rtype: int
        """
        # Pre-process conflicting pairs: map right endpoint to list of left endpoints
        mystic_seals = [[] for _ in range(n + 1)]
        for sigil in conflictingPairs:
            mystic_seals[max(sigil)].append(min(sigil))

        total_lore_fragments = 0
        prime_inhibitor = 0
        secondary_inhibitor = 0

        liberation_scrolls = [0] * (n + 1) # Using list as an array for direct indexing

        # Sweep line from left to right (1 to n)
        for scrying_point in range(1, n + 1):
            for forbidden_origin in mystic_seals[scrying_point]:
                if forbidden_origin > prime_inhibitor:
                    secondary_inhibitor = prime_inhibitor
                    prime_inhibitor = forbidden_origin
                elif forbidden_origin > secondary_inhibitor:
                    secondary_inhibitor = forbidden_origin
            
            # Calculate baseline valid subarrays for current scrying_point
            total_lore_fragments += scrying_point - prime_inhibitor
            
            # Accumulate potential gain if the primary inhibitor were removed
            if prime_inhibitor > 0:
                liberation_scrolls[int(prime_inhibitor)] += prime_inhibitor - secondary_inhibitor

        # Find the maximum possible gain from removing one conflicting pair
        maximum_insight = 0
        for scroll in liberation_scrolls:
            maximum_insight = max(maximum_insight, scroll)

        # Return total baseline + maximum gain
        return total_lore_fragments + maximum_insight
